gum/backend-darwin/gumdarwinsymbolicator.c

```diff
@@ -457,6 +485,123 @@ gum_darwin_symbolicator_find_functions_matching (GumDarwinSymbolicator * self,
 }
 
+gboolean
+gum_darwin_symbolicator_objc_details_from_address (GumDarwinSymbolicator * self,
+                                                   GumAddress address,
+                                                   GumDebugSymbolDetails * details)
+{
+  gboolean success = FALSE;
+  gchar * selector;
+  GumModuleMap * module_map;
+  const GumModuleDetails * module_details;
+  GumDarwinModule * module;
+  GumCollectFunctionsOperation op;
+  GumCollectedFunction dummy_function, * matched_function;
+  GumApiResolver * resolver;
+
+  selector = NULL;
+  module_map = gum_module_map_new ();
+  module_details = gum_module_map_find (module_map, address);
+
+  if (module_details == NULL)
+    goto unref_module_map;
+
+  module = gum_darwin_module_new_from_memory (module_details->path, self->task,
+      module_details->range->base_address, GUM_DARWIN_MODULE_FLAGS_NONE, NULL);
+
+  op.functions = g_array_new (FALSE, FALSE, sizeof (GumCollectedFunction));
+  op.linkedit = module->image->data;
+
+  gum_darwin_module_enumerate_function_starts (module, gum_collect_functions,
+      &op);
+
+  dummy_function.address = address;
+  dummy_function.size = 0;
+
+  matched_function = bsearch (&dummy_function, op.functions->data,
+      op.functions->len, sizeof (GumCollectedFunction),
+      (GCompareFunc)gum_compare_collected_functions);
+
+  resolver = gum_api_resolver_make ("objc");
+
+  _gum_objc_api_resolver_selector_from_address (resolver,
+      matched_function->address, &selector, NULL);
+
+  if (selector == NULL)
+    goto unref_the_rest;
+
+  success = TRUE;
+  details->address = address;
+  g_strlcpy (details->symbol_name, selector, sizeof (details->symbol_name));
+  g_strlcpy (details->module_name, module->name, sizeof (details->module_name));
+
+  g_free (selector);
+
+unref_the_rest:
+  g_object_unref (module);
+  g_array_free (op.functions, TRUE);
+  g_object_unref (resolver);
+
+unref_module_map:
+  g_object_unref (module_map);
+
+  return success;
+}
+
+static gboolean
+gum_collect_functions (const GumDarwinFunctionStartsDetails * details,
+                       gpointer user_data)
+{
+  GumCollectFunctionsOperation * op;
+  const guint8 * p, * end;
+  GumCollectedFunction * prev_function;
+  guint32 offset;
+  GumCollectedFunction function;
+
+  op = user_data;
+  p = (const guint8 *)details->vm_address;
+  end = p + details->size;
+  prev_function = NULL;
+
+  offset = 0;
+  while (p != end)
+  {
+    guint64 delta;
+
+    delta = gum_read_uleb128 (&p, end);
+    if (delta == 0)
+      break;
+
+    if (prev_function != NULL)
+      prev_function->size = delta;
+
+    offset += delta;
+    function.address = GUM_ADDRESS (offset + op->linkedit);
+    prev_function = &function;
+
+    g_array_append_val (op->functions, function);
+  }
+
+  /* The very last function won't have a delta, and so no size, which
+   * means we won't be able to detect mid-function pointers there.
+   */
+  prev_function->size = 0;
```

Comment:
we could assume the size is the rest of the `__text` section in that case, information which could be collected ahead of time alongside the function starts

Reply:
This is done in b5a1381618d3617376bf4ba9d6b0ee003569613a.